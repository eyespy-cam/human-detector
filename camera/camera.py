class Camera:
    def __init__(self):
        self.camera = self.open_camera()

    def open_stream():
        return ""
    
    def open_camera():
        return None

    def get_frame():
        return None

def main():
    pass

if __file__ == "__main__":
    main()