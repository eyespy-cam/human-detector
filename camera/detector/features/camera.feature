Feature: Camera detection

    Scenario: Human is detected
        Given a human walks into frame
        When a human is detected
        Then a notification will be sent

    Scenario: Unauthorized cCamera
        Given a human walks into frame
        When a human is detected
        Then a notification fails to send

    Scenario: No human is detected
        Given a human is not in frame
        When a human is not detected
        Then a notification will not be sent
