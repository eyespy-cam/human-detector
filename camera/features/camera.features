# unauthrized camera
# human
# no human