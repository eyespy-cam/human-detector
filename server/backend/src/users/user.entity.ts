import {
  Collection,
  Entity,
  OneToMany,
  PrimaryKey,
  Property,
  UuidType,
} from '@mikro-orm/core';
import { Group } from '../groups/group.entity';
import { v4 } from 'uuid';

@Entity()
export class User {
  @PrimaryKey({ type: UuidType })
  id = v4();

  // A user might not have a push notification:
  // - After account registration
  // - If they never allow notification permissions in the app
  @Property({ nullable: true })
  expoToken?: string;

  @OneToMany(() => Group, (group) => group.user)
  groups = new Collection<Group>(this);
}
