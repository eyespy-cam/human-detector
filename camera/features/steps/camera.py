from behave import *
import cv2
import os
from src.detector import Detector
from eyeSocket import ServerConnection
from src.camera import Camera

@given('a human walks into frame')
def step_impl(ctx):
    dirname = os.path.realpath('.')
    ctx.frame = cv2.imread(os.path.join(dirname, "tests/images/human.jpg"))
    
@given('a human is not in frame')
def step_impl(ctx):
    dirname = os.path.realpath('.')
    ctx.frame = cv2.imread(os.path.join(dirname, "tests/images/no_human.jpg"))

@when('a human is detected')
def step_impl(ctx):
    detector = Detector()
    result = detector.detect(ctx.frame)
    assert len(result) == 1

@when('a human is not detected')
def step_impl(ctx):
    detector = Detector()
    result = detector.detect(ctx.frame)
    assert len(result) == 0

@then('a notification will be sent')
def step_impl(ctx):
    socket = ServerConnection("ABCD", "https://api.eyespy.com")
    result = socket.send_notification(ctx.frame)
    assert result.status_code == 200

@then('a notification will not be sent')
def step_impl(ctx):
    pass

@then('a notification fails to send')
def step_impl(ctx):
    socket = ServerConnection("AKJFD:LSJF", "https://api.eyespy.com")
    result = socket.send_notification(ctx.frame)
    assert result.status_code != 200
