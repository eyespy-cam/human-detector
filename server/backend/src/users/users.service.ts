import { Collection, EntityRepository } from '@mikro-orm/core';
import { InjectRepository } from '@mikro-orm/nestjs';
import { Injectable } from '@nestjs/common';
import { NotFoundError } from '../errors.types';
import { Camera } from '../cameras/camera.entity'
import { Group } from '../groups/group.entity';
import { User } from './user.entity';

@Injectable()
export class UsersService {
  constructor(
    @InjectRepository(User) private usersRepository: EntityRepository<User>,
  ) {}

  /**
   * Gets the given user's camera groups.
   *
   * @param userId the user's ID.
   * @throws {@link NotFoundError}
   * Thrown if the given user ID doesn't correspond to an existing user.
   */
  public async getGroups(userId: string): Promise<Collection<Group>> {
    const user = await this.usersRepository.findOne(
      { id: userId },
      { populate: ['groups.cameras'] },
    );
    if (user === null) {
      throw new NotFoundError(`User with ID "${userId}" does not exist.`);
    }
    return user.groups;
  }

  public async putCamera(userId: string, groupId: string, name: string, pubKey: string, serial: string): Promise<Camera> {
    
    return null;
  }

  /**
   * Update's the user's push notification token.
   *
   * @param userId the user's ID.
   * @param notifyToken the new notification token.
   */
  public async updateNotifyToken(
    userId: string,
    notifyToken: string,
  ): Promise<void> {
    const user = await this.usersRepository.findOne({ id: userId });
    if (user === null) {
      throw new NotFoundError(`User with ID "${userId}" does not exist.`);
    }
    user.expoToken = notifyToken;
    await this.usersRepository.flush();
  }
}
