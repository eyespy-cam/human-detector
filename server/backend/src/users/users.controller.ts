import {
  Body,
  Controller,
  ForbiddenException,
  Get,
  Inject,
  Param,
  Put,
  UseGuards,
} from '@nestjs/common';
import { JwtIdentityGuard } from '../auth/jwt-identity.guard';
import { NotFoundError } from '../errors.types';
import { UsersService } from './users.service';

export type GetGroupsOutput = {
  id: string;
  name: string;
  cameras: {
    id: string;
    name: string;
  }[];
}[];

@Controller('users')
@UseGuards(JwtIdentityGuard)
export class UsersController {
  constructor(@Inject(UsersService) private usersService: UsersService) {}

  @Get(':id/groups')
  async getGroups(@Param('id') id: string): Promise<GetGroupsOutput> {
    try {
      const groupsJSON = (await this.usersService.getGroups(id)).toJSON();
      return groupsJSON.map((group) => ({
        id: group.id,
        name: group.name,
        cameras: group.cameras.map((camera) => ({
          id: camera.id,
          name: camera.name,
        })),
      }));
    } catch (error) {
      if (error instanceof NotFoundError) {
        throw new ForbiddenException();
      } else {
        throw error;
      }
    }
  }

  @Put(':id/notifyToken')
  async updateNotifyToken(
    @Param('id') id: string,
    @Body('expoToken') notifyToken: string,
  ): Promise<void> {
    try {
      await this.usersService.updateNotifyToken(id, notifyToken);
    } catch (error) {
      if (error instanceof NotFoundError) {
        throw new ForbiddenException();
      } else {
        throw error;
      }
    }
  }
}
