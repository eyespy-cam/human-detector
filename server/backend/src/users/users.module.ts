import { MikroOrmModule } from '@mikro-orm/nestjs';
import { Module } from '@nestjs/common';
import { Group } from '../groups/group.entity';
import { User } from './user.entity';
import { UsersController } from './users.controller';
import { UsersService } from './users.service';

@Module({
  imports: [MikroOrmModule.forFeature([User, Group])],
  controllers: [UsersController],
  providers: [UsersService],
})
export class UsersModule {}
