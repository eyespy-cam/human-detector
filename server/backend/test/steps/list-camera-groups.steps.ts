import { INestApplication } from '@nestjs/common';
import { defineFeature, loadFeature } from 'jest-cucumber';
import { v4 } from 'uuid';
import request from 'supertest';
import { EntityRepository, MikroORM } from '@mikro-orm/core';
import { User } from '../../src/users/user.entity';
import { Group } from '../../src/groups/group.entity';
import { Camera } from '../../src/cameras/camera.entity';
import { GetGroupsOutput } from '../../src/users/users.controller';
import {
  buildTestStack,
  TestStack,
  TEST_STACK_TIMEOUT,
} from '../helpers/test-stack';
import { UsersModule } from '../../src/users/users.module';

defineFeature(
  loadFeature('test/features/list-camera-groups.feature'),
  (test) => {
    let app: INestApplication;
    let testStack: TestStack;
    let userRepository: EntityRepository<User>;

    beforeAll(async () => {
      testStack = await buildTestStack({ imports: [UsersModule] });
      userRepository = testStack.module
        .get<MikroORM>(MikroORM)
        .em.getRepository(User);

      app = testStack.module.createNestApplication();
      await app.init();
    }, TEST_STACK_TIMEOUT);

    afterAll(async () => {
      await app.close();
      await testStack.kcContainer.stop();
      await testStack.dbContainer.stop();
    });

    test('User is viewing their camera groups from the app', ({
      given,
      when,
      then,
    }) => {
      let user: User;
      let token: string;
      let response: request.Response;

      given('I have credentials', async () => {
        const { id, tokenSet } =
          await testStack.kcContainer.createDummyUserAndLogIn('users');
        user = await userRepository.findOneOrFail({ id });
        user.groups.add(new Group('group-a'));
        user.groups[0].cameras.add(
          new Camera('My camera :)', 'wajebawk', 'definitely a serial'),
        );
        await userRepository.flush();
        token = tokenSet.access_token;
      });
      when('I request a list of my camera groups', async () => {
        response = await request(app.getHttpServer())
          .get(`/users/${user.id}/groups`)
          .auth(token, { type: 'bearer' });
      });
      then(
        'I receive my camera groups, including the list of cameras associated with each group',
        () => {
          const expectedResponse: GetGroupsOutput = [
            {
              id: user.groups[0].id,
              name: user.groups[0].name,
              cameras: [
                {
                  id: user.groups[0].cameras[0].id,
                  name: user.groups[0].cameras[0].name,
                },
              ],
            },
          ];
          expect(response.statusCode).toBe(200);
          expect(response.type).toEqual('application/json');
          expect(response.body).toEqual(expectedResponse);
        },
      );
    });

    test("User A is attempting to list another user B's camera groups", ({
      given,
      when,
      then,
    }) => {
      let userB: User;
      let userTokenA: string;
      let response: request.Response;

      given("I have user A's credentials", async () => {
        const userBLogin = await testStack.kcContainer.createDummyUserAndLogIn(
          'users',
        );
        userB = await userRepository.findOneOrFail({ id: userBLogin.id });
        const userALogin = await testStack.kcContainer.createDummyUserAndLogIn(
          'users',
        );
        userTokenA = userALogin.tokenSet.access_token;
      });
      when("I request user B's camera groups", async () => {
        response = await request(app.getHttpServer())
          .get(`/users/${userB.id}/groups`)
          .auth(userTokenA, { type: 'bearer' });
      });
      then('I receive an unauthorized error', () => {
        expect(response.statusCode).toBe(403);
        expect(response.body).toMatchSnapshot();
      });
    });

    test('User A is attempting to list the camera groups of user B, a non-existent user', ({
      given,
      when,
      then,
    }) => {
      const userIdB = v4();
      let userTokenA: string;
      let response: request.Response;

      given("I have user A's credentials", async () => {
        const { tokenSet } =
          await testStack.kcContainer.createDummyUserAndLogIn('users');
        userTokenA = tokenSet.access_token;
      });
      when("I request user B's camera groups", async () => {
        response = await request(app.getHttpServer())
          .get(`/users/${userIdB}/groups`)
          .auth(userTokenA, { type: 'bearer' });
      });
      then('I receive an unauthorized error', () => {
        expect(response.statusCode).toBe(403);
        expect(response.body).toMatchSnapshot();
      });
    });

    test('User is attempting to view their camera groups with an expired token', ({
      given,
      when,
      then,
    }) => {
      const userId = v4();
      let token: string;
      let response: request.Response;

      given("I have user A's expired credentials", async () => {
        const { id, tokenSet } =
          await testStack.kcContainer.createDummyUserAndLogIn('users');
        token = tokenSet.access_token;

        // This seems to be the only way to cause a user's credentials to expire without
        // waiting the entire expiration duration.
        await testStack.kcContainer.logUserOut('users', id);
      });
      when("I request user A's camera groups", async () => {
        response = await request(app.getHttpServer())
          .get(`/users/${userId}/groups`)
          .auth(token, { type: 'bearer' });
      });
      then('I receive an unauthorized error', () => {
        expect(response.statusCode).toBe(403);
        expect(response.body).toMatchSnapshot();
      });
    });
  },
);
