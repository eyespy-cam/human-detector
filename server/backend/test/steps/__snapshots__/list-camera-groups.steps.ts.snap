// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Listing a user's camera groups User A is attempting to list another user B's camera groups 1`] = `
Object {
  "message": "Forbidden",
  "statusCode": 403,
}
`;

exports[`Listing a user's camera groups User A is attempting to list the camera groups of user B, a non-existent user 1`] = `
Object {
  "message": "Forbidden",
  "statusCode": 403,
}
`;

exports[`Listing a user's camera groups User is attempting to view their camera groups with an expired token 1`] = `
Object {
  "message": "Unauthorized",
  "statusCode": 401,
}
`;
